test/gtest/ucp/test_ucp_atomic.cc

```diff
@@ -153,17 +198,67 @@ class test_ucp_atomic : public test_ucp_memheap {
         return request_wait(status_ptr);
     }
 
+    void test_mem_types(send_func_t send_func, unsigned num_iters,
+                        uint64_t op_mask, int is_ep_flush) {
+        const char *file = "/sys/kernel/mm/memory_peers/nv_mem/version";
+        int atomic_mode = get_variant_value() & ATOMIC_MODE;
+        std::vector<std::vector<ucs_memory_type_t> > pairs =
+                ucs::supported_mem_type_pairs();
+        bool have_peerdirect = !access(file, F_OK);
+
+        for (size_t i = 0; i < pairs.size(); ++i) {
+            if (!UCP_MEM_IS_HOST(pairs[i][0]) || !UCP_MEM_IS_HOST(pairs[i][1])) {
+                /* Memory type atomics are fully supported only with new protocols */
+                if (!(get_variant_value() & ENABLE_PROTO)) {
+                    continue;
+                }
+
+                static const std::string tls[] = { "ud_v", "ud_x", "rc_v", "tcp" };
+                /* Target memory type atomics emulation not supported yet */
+                if ((atomic_mode == UCP_ATOMIC_MODE_CPU ||
+                     has_any_transport(std::vector<std::string>(tls,
+                             tls + ucs_static_array_size(tls)))) &&
```

Comment:
can you initialize the vector from start,size instead of start,end? (avoid "tls+")

Reply:
has_any_transport to accept C-array, size